block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -75,6 +76,10 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
     def isFinalized(blockHash: BlockHash): F[Boolean] =
       finalizedBlocksSet.contains(blockHash).pure[F]
 
+    override def find(truncatedHash: String): F[Option[BlockHash]] = Sync[F].delay {
+      dagSet.find(hash => Base16.encode(hash.toByteArray).startsWith(truncatedHash))
```

Comment:
I'm curious, would something like this work?
```scala
 override def find(truncatedHash: String): F[Option[BlockHash]] = Sync[F].delay {
      dagSet.find { hash =>
        // Use iterators to make it as lazy as possible, and transform only short string, not the whole hash
        val searchPrefix = Base16.unsafeDecode(truncatedHash).toIterator
        // This fails fast on the first array element mismatch
        !(hash.toByteArray.toIterator zip searchPrefix).exists { case (l, r) => l != r }
      }
    }
```

Reply:
@nzpr If you suspecting that bottleneck is in byte comparison but I think this is already optimized in Java `String` class in `startWith` method.

I'm more suspicious in our custom implementation of encoding bytes to hex. This is used in `Base16.encode`.
https://github.com/rchain/rchain/blob/57eb112cf7275b31316b4692d2faf0f62fab0dad/crypto/src/main/scala/coop/rchain/crypto/codec/Base16.scala#L32-L33

Reply:
I see, anyway as it was shown by @stanislavlyalin on standup, there is `startsWith` on ByteString, so the fastest way is to convert `truncatedHash` into ByteString